### PR TITLE
More than one partition

### PR DESCRIPTION
**Pull request by jgrillo-grapl**
### Which issue does this PR correspond to?
https://app.zenhub.com/workspaces/grapl-6036cbd36bacff000ef314f2/issues/grapl-security/issue-tracker/992

### What changes does this PR make to Grapl? Why?

This PR updates the local Grapl deployment to operate Kafka under a better approximation of realistic conditions. In particular, we now have a cluster of 3 brokers, and we set `min.insync.replicas=2`, `partitions=2`, and `replication=3` on each topic. Additionally, we now run 2 instances of each consumer service.

Also, we're now committing our offsets in the `kafka` library. Also, before we had set `auto.offset.reset="earliest"`, now we use `"latest"`. I've also made a bunch of updates to the `KafkaTopicScanner`:

1. Now we have a unique consumer group for every `KafkaTopicScanner` instance. This ensures that each scanner consumes messages from every partition.
2. The `KafkaTopicScanner#contains*` methods return a `JoinHandle<Vec<Envelope<T>>>` containing _all_ the messages matching the predicate instead of just the first one. This makes the tool a little more versatile because now we can make assertions about things like the number of matching messages and interrelationships between messages. Also, the `KafkaTopicScanner` does no error handling internally. Every error panics. These panics are propagated to the test function so we can see useful things like the panic message, stack traces, etc.
3. The `KafkaTopicScanner` now commits its offsets.

### How were these changes tested?
Automated tests

### CODE REVIEW

File: nomad/local/grapl-local-infra.nomad
Comment (wimax-grapl) [inline nomad/local/grapl-local-infra.nomad:6]: I think I'd much prefer a 
```
variable "kafka_broker_args" {
  type = list(object({
    broker_port_listen = number
    broker_port_host_os = number
    jmx_port = number
  }))
}
```

but I won't push the matter;  notably this would require you constructing that arg as json in startdetach

File: nomad/local/grapl-local-infra.nomad
Comment (wimax-grapl) [inline nomad/local/grapl-local-infra.nomad:273]: is this ever used?

File: nomad/local/grapl-local-infra.nomad
Comment (wimax-grapl) [inline nomad/local/grapl-local-infra.nomad:323]: seeing how much is repeated here I'd really prefer if we used a `dynamic "task" {}` which is literally perfect for these purposes. it iterates over the aforementioned `variable "kafka_broker_args"` and you can inject `i.broker_port`, `i.jmx_port` etc.

Comment (wimax-grapl) [reply nomad/local/grapl-local-infra.nomad:323]: see `    for_each = local.database_descriptors` for an example

File: nomad/local/grapl-local-infra.nomad
Comment (wimax-grapl) [inline nomad/local/grapl-local-infra.nomad:559]: remove so we can see if the other postgres fix did the trick?

File: src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs
Comment (wimax-grapl) [inline src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs:1]: this seems wrong

File: src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs
Comment (wimax-grapl) [inline src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs:72]: can we bring back my predicates?

Comment (wimax-grapl) [reply src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs:72]: oh i see they're done below in-line, which i'm happy about; but i'm a bit confused about the role of the predicate here now if we do filtering later on

File: src/rust/kafka/src/lib.rs
Comment (wimax-grapl) [inline src/rust/kafka/src/lib.rs:540]: thank you for the additional docs

File: nomad/local/grapl-local-infra.nomad
Comment (jgrillo-grapl) [reply nomad/local/grapl-local-infra.nomad:6]: Oh awesome, thanks! I'll do that and use a dynamic task.

File: src/rust/kafka/src/lib.rs
Comment (wimax-grapl) [inline src/rust/kafka/src/lib.rs:592]: i think `result.and_then(|| self.consumer.commit())` would do the trick but it's fine as is

File: src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs
Comment (jgrillo-grapl) [reply src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs:72]: Yeah, the idea is to just do _filtering_ in the `contains`, then make assertions on the filtered results. It just makes it a lot easier to tell what went wrong when tests fail.

I think maybe I should change the `contains` names to be `filter`, actually. Does that sound OK?

File: src/rust/kafka/src/test_utils/topic_scanner.rs
Comment (wimax-grapl) [inline src/rust/kafka/src/test_utils/topic_scanner.rs:108]: gonna skip reviewing this part re: rx-tx, but the stop/filter predicate _is_ cool

Comment (wimax-grapl) [reply src/rust/kafka/src/test_utils/topic_scanner.rs:108]: i guess `contains` is no longer the right term though, maybe just `scan` or `filter`. we can fix that up later though

File: nomad/local/grapl-local-infra.nomad
Comment (jgrillo-grapl) [reply nomad/local/grapl-local-infra.nomad:273]: Yeah it's used in the healthcheck on L309.

File: src/rust/node-identifier/tests/integration_test.rs
Comment (wimax-grapl) [inline src/rust/node-identifier/tests/integration_test.rs:116]: this is so much more ergonomic :+1:

File: src/rust/pipeline-ingress/tests/integration_test.rs
Comment (wimax-grapl) [inline src/rust/pipeline-ingress/tests/integration_test.rs:131]: doesn't the span fill in those two things for us already?

Comment (jgrillo-grapl) [reply src/rust/pipeline-ingress/tests/integration_test.rs:131]: The span only exists in places where we've consumed an Envelope<T> from Kafka. It's automatically applied to everything that runs in the StreamProcessor's closure, but that's about it. Here we don't have it yet because we don't consume anything (that happens in the topic scanner).

File: nomad/local/grapl-local-infra.nomad
Comment (jgrillo-grapl) [reply nomad/local/grapl-local-infra.nomad]: done

File: nomad/local/grapl-local-infra.nomad
Comment (jgrillo-grapl) [reply nomad/local/grapl-local-infra.nomad]: done

File: nomad/local/grapl-local-infra.nomad
Comment (jgrillo-grapl) [reply nomad/local/grapl-local-infra.nomad]: done, still seeing issues that appear to be caused by lack of postgres memory. In particular, seeing this repeatedly in the `plugin-work-queue-db` error logs: 
```
2022-08-08 17:49:46.642 UTC [1] LOG:  server process (PID 111) was terminated by signal 9: Killed
```

File: src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs
Comment (jgrillo-grapl) [reply src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs]: done

File: src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs
Comment (jgrillo-grapl) [reply src/rust/e2e-tests/tests/sysmon_log_e2e_test.rs]: done. Changed them to `scan` and `scan_for_tenant`

File: src/rust/kafka/src/lib.rs
Comment (jgrillo-grapl) [reply src/rust/kafka/src/lib.rs:586]: done

File: src/rust/kafka/src/test_utils/topic_scanner.rs
Comment (jgrillo-grapl) [reply src/rust/kafka/src/test_utils/topic_scanner.rs:108]: fixed

File: src/rust/node-identifier/tests/integration_test.rs
Comment (jgrillo-grapl) [reply src/rust/node-identifier/tests/integration_test.rs:116]: thanks!

**Review by wimax-grapl — APPROVED** [review]
approved assuming we can get the postgres stuff working

File: src/rust/kafka/src/lib.rs
Comment (wimax-grapl) [inline src/rust/kafka/src/lib.rs:581]: I'd imagine you could remove the Ok() and the ?

Comment (wimax-grapl) [reply src/rust/kafka/src/lib.rs:581]: but if not, whatever

Comment (wimax-grapl) [reply src/rust/kafka/src/lib.rs:581]: maybe the ? is doing some conversion for us

**Review by wimax-grapl — APPROVED** [review]

File: src/rust/kafka/src/lib.rs
Comment (jgrillo-grapl) [reply src/rust/kafka/src/lib.rs:581]: Yes, the `?` is converting the `CommitError` for us

File: nomad/local/nomad-agent-conf.nomad
Comment (twunderlich-grapl) [inline nomad/local/nomad-agent-conf.nomad:45]: We're tuning nomad gc to prevent logs from being collected, right? Might be nice to put a comment pertaining to that above.

Comment (jgrillo-grapl) [reply nomad/local/nomad-agent-conf.nomad:45]: Yeah, will do

Comment (jgrillo-grapl) [reply nomad/local/nomad-agent-conf.nomad:47]: done

**Review by ghost — APPROVED** [review]
Approving the Nomad, postgres config and pulumi code. I'll take a quick scan through the rust code, but I imagine that you'll get better reviews from Max et al

**Pull request merged**
